L1Trigger/L1TGEM/src/GE0TriggerPseudoBuilder.cc

```diff
@@ -0,0 +1,164 @@
+  auto detid = segment.gemDetId();
+  const GEMSuperChamber* chamber = me0_g->superChamber(detid);
+  const GEMChamber* keylayer = chamber->chamber(GE0TriggerPseudoBuilder::ME0KeyLayer);
+  int chamberid = detid.superChamberId() % 2;
+  int totRolls = keylayer->nEtaPartitions();
+  float dphi = chamber->computeDeltaPhi(segment.localPosition(), segment.localDirection());
+  // !!!!TODO!!!! float time = segment.time();
+  // !!!!TODO!!!! int sign_time = (time > 0) ? 1 : -1;
+  int nrechits = segment.nRecHits();
+  std::vector<int> rolls;
+  for (const auto& rechit : segment.specificRecHits()) {
+    if (std::find(rolls.begin(), rolls.end(), rechit.gemId().roll()) == rolls.end())
+      rolls.push_back(rechit.gemId().roll());
+  }
+  if (rolls.size() > 2 or rolls.empty())
+    LogTrace("L1ME0Trigger") << " ME0 segment is crossing " << rolls.size() << " roll !!! \n";
+  //assert(rolls.size() <= 2);   // we did found very few ME0 segments crossing 3 rolls!!! this cut is applied offline
+  if (rolls.empty())
+    return ME0TriggerDigi();
+  if (rolls[0] < 1)
+    LogTrace("L1ME0Trigger") << " ME0 segment has wrong roll number " << rolls[0] << " which should be >= 1 \n !!!";
+  assert(rolls[0] >= 1);
+  int partition = (rolls[0] - 1) << 1;  //roll from detid counts from 1
+  if (rolls.size() == 2 and rolls[0] > rolls[1])
+    partition = partition - 1;
+  else if (rolls.size() == 2 and rolls[0] < rolls[1])
+    partition = partition + 1;
+
+  if (partition < 0 or partition >= 2 * totRolls) {
+    LogTrace("L1ME0Trigger") << " ME0 segment rolls size of all hits " << rolls.size() << " rolls[0] " << rolls[0]
+                             << " rolls.back() " << rolls.back() << " and ME0 trigger roll is " << partition
+                             << " max expected " << 2 * totRolls - 1 << "\n";
+    return ME0TriggerDigi();
+  }
+
+  //globalpoint from ME0 segment
+  GlobalPoint gp = me0_g->idToDet(segment.gemDetId())->surface().toGlobal(segment.localPosition());
+  const GEMEtaPartition* etapart = keylayer->etaPartition(rolls[0]);
+  LocalPoint segment_lp = etapart->surface().toLocal(gp);  // convert segment gp into lp in etapartition coordinate
+  float strippitch = etapart->localPitch(segment_lp);
+  float strip = etapart->strip(segment_lp);
+  int totstrip = etapart->nstrips();
+  int istrip = static_cast<int>(strip);
+  int phiposition = istrip;
+  if (phiposition > totstrip)
+    LogTrace("L1ME0Trigger") << " ME0 segment strip number is " << phiposition << " larger than nstrip " << totstrip
+                             << " !!! \n";
+  float phi_resolution = 0.5;                                                         //halfstrip
+  int phiposition2 = (static_cast<int>((strip - phiposition) / phi_resolution) & 1);  // half-strip resolution
+  phiposition = (phiposition << 1) | phiposition2;
+
+  //gloablpoint from ME0 trigger digi
+  float centreOfStrip = istrip + 0.25 + phiposition2 * 0.5;
```

Comment:
My understanding is the following:

1.  Doing this calculation in half-strips:
   - `centerOfStrip2 = istrip * 2 + 0.5 + phiposition2`
   - Note that the 0.5 in this case indicates a quarter of a half-strip since that's the resolution they're using.
2. Divide by 2 to get the strip number:
   - `centerOfStrip = istrip + 0.25 + phiposition2 * 0.5`
   - This would mean the `0.25` is a quarter of a strip, and the `0.5` is the phi resolution.